### ISSUE TITLE
Different weight for each class in training loss

### ISSUE THREAD

**Issue by ByoenghakYim** [weiliu89/caffe#291]
Hi

I would like to know how modify multibox_loss_layer.cpp to assign the different weight for each class during training time.

I think that I have to modify smooth_L1_loss_layer & softmax_layer also.

But to modify them, I need to know whole data structure they use.

So I was wondering if you have any idea to implement it or any hint of which variable has "Label" for each ground truth.

Thanks in advance

**Comment by weiliu89** [comment]
If you read the code, multibox_loss_layer.cpp created two loss layers, one for localization loss and the other for confidence loss. If you want to add weight to them, you could look into [here](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/multibox_loss_layer.cpp#L286-L335) for how I form the input for localization loss, and [here](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/multibox_loss_layer.cpp#L375-L436) for how I form the input for confidence loss.

**Comment by ByoenghakYim** [comment]
thanks for ur reply.

I'll inform you after I

**Comment by ByoenghakYim** [comment]
thanks for ur reply.

I'll inform you after trying

**Comment by Goldit** [comment]
@ByoenghakYim,  have you tried it yet? 
I would also like to assign different weight for different classes to take into account class imbalance.

**Comment by ByoenghakYim** [comment]
@Goldit 
I've tried but the result is not different from the original.

**Comment by JackieLeeTHU11** [comment]
@ByoenghakYim Why you want to try the class weight for different class? I want the train ssd with my dataset. However, the classes are imblance. So I want to try the class weight for different classes. Have you check in the modified multibox_loss_layer.cpp in Github or somewhere else?

**Comment by ByoenghakYim** [comment]
@JackieLeeTHU11 
I'm still on validation.
With VGGNet, it show some improvements, 3~10 percentage on KITTI.
But more data shows better performance because weighted learning rate is just like sampling same data more.
You need to modify multibox_loss_layer & bbox_util & loss function code

**Comment by JackieLeeTHU11** [comment]
@ByoenghakYim 
Thank you. I read the multibox_loss_layer. The author @weiliu89 utilize a SoftmaxWithLoss layer in the conf loss. So I think we can only modify the SoftmaxWithLoss layer and add class weights in the layer.
